dbms/src/AggregateFunctions/AggregateFunctionEntropy.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <AggregateFunctions/FactoryHelpers.h>
+
+/// These must be exposed in header for the purpose of dynamic compilation.
+#include <AggregateFunctions/QuantileReservoirSampler.h>
+#include <AggregateFunctions/QuantileReservoirSamplerDeterministic.h>
+#include <AggregateFunctions/QuantileExact.h>
+#include <AggregateFunctions/QuantileExactWeighted.h>
+#include <AggregateFunctions/QuantileTiming.h>
+#include <AggregateFunctions/QuantileTDigest.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/QuantilesCommon.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#include <type_traits>
+
+#include <cmath>
+#include <AggregateFunctions/UniqVariadicHash.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+
+/** Calculates Shannon Entropy, using HashMap and computing empirical distribution function
+  */
+template <typename Value>
+struct EntropyData
+{
+    using Weight = UInt64;
+    using Map = HashMap <
+    Value, Weight,
+    HashCRC32<Value>,
```

Comment:
If the value is already hashed (UInt128), we should use TrivialHash instead of HashCRC32.